data/org.freedesktop.portal.DynamicLauncher.xml

```diff
@@ -0,0 +1,287 @@
+        "org.gnome.Epiphany.WebApp_e9d0e1e4b0a10856aa3b38d9eb4375de4070d043.desktop"
+        In that example the desktop file would exist at the path
+        "~/.local/share/xdg-desktop-portal/applications/org.gnome.Epiphany.WebApp_e9d0e1e4b0a10856aa3b38d9eb4375de4070d043.desktop"
+        with a sym link in "~/.local/share/applications/".
+        The checksum at the end of the file name is an implementation detail in
+        Epiphany and not required by the portal.
+
+        If the calling process is unsandboxed, @desktop_file_id need only have
+        a ".desktop" suffix.
+
+        This method is intended to be called by the application that created the
+        launcher, e.g. a web browser, so it can clean up associated data such as
+        the profile directory as part of the uninstallation. Consequently, the
+        proper way for a software center to remove a launcher is by using the APIs
+        provided by the application that installed it. For example, for GNOME
+        Software to remove web launchers created by Epiphany, it would use the
+        org.gnome.Epiphany.WebAppProvider D-Bus interface.
+
+        Please note that this method call will fail if the specified launcher
+        already does not exist.
+
+        The @options vardict currently has no supported entries.
+    -->
+    <method name="Uninstall">
+      <arg type="s" name="desktop_file_id" direction="in"/>
+      <arg type="a{sv}" name="options" direction="in"/>
+    </method>
+    <!--
+        GetDesktopEntry:
+        @desktop_file_id: The .desktop file name
+        @contents: the contents of the named .desktop file
+
+        This function returns the contents of a desktop file with the name
+        @desktop_file_id in @contents. If the calling process is sandboxed,
+        @desktop_file_id must have the app ID followed by a period as a prefix;
+        see the Uninstall() method documentation.
+
+        If the calling process is unsandboxed, @desktop_file_id need only have
+        a ".desktop" suffix.
+
+        This method only works for desktop files that were created by the
+        dynamic launcher portal.
+    -->
+    <method name="GetDesktopEntry">
+      <arg type="s" name="desktop_file_id" direction="in"/>
+      <arg type="s" name="contents" direction="out"/>
+    </method>
+    <!--
+        Launch:
+        @desktop_file_id: The .desktop file name
+
+        This function launches the app specified by @desktop_file_id. If the
+        calling process is sandboxed, @desktop_file_id must have the app ID
+        followed by a period as a prefix; see the Uninstall() method
+        documentation.
+
+        If the calling process is unsandboxed, @desktop_file_id need only have
+        a ".desktop" suffix.
+
+        This method only works for desktop files that were created by the
+        dynamic launcher portal.
+    -->
+    <method name="Launch">
+      <arg type="s" name="desktop_file_id" direction="in"/>
+    </method>
```

Comment:
This API is missing two things that might be considered important:

- a timestamp, which per the docs of gdk_app_launch_context_set_timestamp():

> The timestamp should ideally be taken from the event that triggered the launch.
> 
> Window managers can use this information to avoid moving the focus to the newly launched application when the user is busy typing in another window. This is also known as ‘focus stealing prevention’.

- and a way to set what screen the launch happens on, which is not needed by Epiphany but which could be needed in some contexts. However, on Wayland it is [not possible](https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/4434/diffs) to control the workspace or screen a window is launched on (according to ebassi). And on X11 it is only possible to control the workspace (via `gdk_app_launch_context_set_desktop()` not the physical monitor.

So perhaps a way to control the workspace is not worth the trouble, and just a timestamp is enough. But it does come at the cost of requiring each backend to implement another method.

Reply:
On closer inspection it looks like there's already a portal that uses `g_app_info_launch`, the open-uri portal. And there we also have support for xdg activation tokens (https://github.com/flatpak/xdg-desktop-portal/pull/566) which appear to be a mechanism for focus stealing prevention that is intended to replace timestamps ([docs](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/50/diffs#diff-content-7b1a427b51eea37983e26cee3b13856c7731eb35)). So we should follow suit in this portal.